### PR TITLE
First Attempt at stack API changes

### PR DESCRIPTION
**Pull request by undertheflowerpot**
Thanks for contributing to this project!

This repository is still work in progress. We made this repository public so that
we can discuss the design and progress with the community, but are still putting 
things in shape.

With things moving fast, we may not be able to review, or accept your pull-request
at this time (but we'll try to have a look - bear with us 🤗).

Thank you for your understanding!

### CODE REVIEW

File: pkg/controller/backend/backend.go [hunk 1]
```diff
@@ -32,120 +26,80 @@ func NewDefaultStacksBackend(stackStore interfaces.StackStore, swarmBackend inte
...
-	if create.Orchestrator != types.OrchestratorSwarm {
-		return types.StackCreateResponse{}, fmt.Errorf("invalid orchestrator type %s. This backend only supports orchestrator type swarm", create.Orchestrator)
-	}
-
-	// Create the Swarm Stack object
-	stack := types.Stack{
-		Spec:         create.Spec,
-		Orchestrator: types.OrchestratorSwarm,
-	}
+func (b *DefaultStacksBackend) CreateStack(stackSpec interfaces.StackSpec) (string, error) {
```
Comment (dperny) [inline pkg/controller/backend/backend.go]: the return value of this ought to be changed back to a `StackCreateResponse`.

File: pkg/interfaces/types.go [hunk 2]
```diff
@@ -12,21 +12,22 @@ const (
...
 // API types.
-// NOTE: SwarmStack is only used internally for reconciliation of Swarm
+// NOTE: Stack is only used internally for reconciliation of Swarm
 // stacks and is never exposed via the API.
-type SwarmStack struct {
-	ID   string
-	Meta swarm.Meta
-	Spec SwarmStackSpec
+type Stack struct {
+	Version swarm.Version
```
Comment (dperny) [inline pkg/interfaces/types.go]: Version is not needed here -- it should be part of Meta

File: pkg/interfaces/types.go [hunk 3]
```diff
@@ -12,21 +12,22 @@ const (
...
+	ID      string
+	Meta    swarm.Meta
```
Comment (dperny) [inline pkg/interfaces/types.go]: This is my mistake, but Meta should actually be an embedded type. That is, the struct should look like this:

```
type Stack struct {
    ID string
    swarm.Meta
    Spec StackSpec
}
```

This keeps with the pattern established with Swarm objects in the existing Docker API

(see here: https://github.com/moby/moby/blob/8d1309222c58143c1e7fa02658e04d06e2e934ce/api/types/swarm/service.go#L5-L13 )

File: pkg/interfaces/types.go [hunk 4]
```diff
@@ -12,21 +12,22 @@ const (
...
+	Spec    StackSpec
 }
 
-// SwarmStackSpec represents a StackSpec with all of its elements converted to
+// StackSpec represents a StackSpec with all of its elements converted to
 // Engine API types.
-// NOTE: SwarmStackSpec is only used internally for reconciliation of
+// NOTE: StackSpec is only used internally for reconciliation of
```
Comment (dperny) [inline pkg/interfaces/types.go]: you can strike this line and the next one. StackSpec is used externally now.

File: pkg/store/marshal.go [hunk 5]
```diff
@@ -10,66 +10,52 @@ import (
...
 // swarmkit and contained in the Resource
-func UnmarshalStacks(resource *api.Resource) (*types.Stack, *interfaces.SwarmStack, error) {
+func UnmarshalStacks(resource *api.Resource) (*interfaces.Stack, error) {
 	iface, err := typeurl.UnmarshalAny(resource.Payload)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 	// this is a naked cast, which means if for some reason this _isn't_ a
 	// CombinedStack object, the program will panic. This is fine, because if
```
Comment (dperny) [inline pkg/store/marshal.go]: You can update this comment; there is no `CombinedStack` anymore.

File: pkg/interfaces/types.go (hunk 2 above)
Comment (undertheflowerpot) [reply pkg/interfaces/types.go]: fixed yet moved to pkg/types/stack.go

**Comment by thaJeztah** [comment]
Looks like the commits are missing a DCO sign-off

File: pkg/controller/backend/backend.go [hunk 6]
```diff
@@ -73,79 +57,46 @@ func (b *DefaultStacksBackend) GetStack(id string) (types.Stack, error) {
...
-func (b *DefaultStacksBackend) convertToSwarmStackSpec(spec types.StackSpec) (interfaces.SwarmStackSpec, error) {
-	// Substitute variables with desired property values
-	substitutedSpec, err := substitution.DoSubstitution(spec)
-	if err != nil {
-		return interfaces.SwarmStackSpec{}, err
-	}
-
+/*
+// FIXME:  DELETE
+func (b *DefaultStacksBackend) convertToSwarmStackSpec(spec types.StackSpec) (types.StackSpec, error) {
```
Comment (dperny) [inline pkg/controller/backend/backend.go]: This still needs to be deleted.

File: pkg/types/stack.go [hunk 7]
```diff
@@ -1,20 +1,41 @@
...
+	StackEventType = "stack"
+	// StackLabel is a label on objects indicating the stack that it belongs to
+	StackLabel = "com.docker.stacks.stack_id"
+)
 
-	// TODO - temporary (not in swagger)
+// Stack represents a Stack with all of its elements converted to Engine
+// API types.
+// NOTE: Stack is only used internally for reconciliation of Swarm
+// stacks and is never exposed via the API.
```
Comment (dperny) [inline pkg/types/stack.go]: We should rewrite this comment, because it reflects the old understanding of "stack"

File: pkg/types/stack.go [hunk 8]
```diff
@@ -1,20 +1,41 @@
...
+type Stack struct {
 	ID string
+	swarm.Meta
+	Spec StackSpec
+}
+
+// StackSpec represents a StackSpec with all of its elements converted to
+// Engine API types.
```
Comment (dperny) [inline pkg/types/stack.go]: Likewise

File: Makefile [hunk 9]
```diff
@@ -5,10 +5,10 @@ E2E_IMAGE_NAME=stack-e2e
 TAG=latest # TODO work out versioning scheme
 TEST_SCOPE?=./...
 BUILD_ARGS= \
-    --build-arg ALPINE_BASE=alpine:3.8 \
-    --build-arg GOLANG_BASE=golang:1.11-alpine3.8
+    --build-arg ALPINE_BASE=alpine:3.10 \
+    --build-arg GOLANG_BASE=golang:1.12.6-alpine3.10
 
-build: generate
+build:
```
Comment (dperny) [inline Makefile]: We can additionally remove the "generate" make rule, as it's no longer needed.

**Pull request merged**